JDK-8313781: Add regression tests for large page logging and user-facing error messages

tstuefe opened the pull request:
In preparation for the UseSHM removal (https://bugs.openjdk.org/browse/JDK-8310841) I'd like more regression tests. 

The new test basically checks the behavior of the JVM when confronted with different variations of:
- OS capabilities
- JVM settings.

For example: If the JVM is started with -XX:+UseLargePages, but on the OS side there are no huge pages configured, or the kernel does not even support hugepages, the JVM should disable large pages and issue a warning.

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8313781](https://bugs.openjdk.org/browse/JDK-8313781): Add regression tests for large page logging and user-facing error messages (**Bug** - P4)


### Reviewers
 * [Stefan Johansson](https://openjdk.org/census#sjohanss) (@kstefanj - **Reviewer**) ⚠️ Review applies to [8cc04b03](https://git.openjdk.org/jdk/pull/15159/files/8cc04b03952fa43b6096d465d205ee2854a30676)
 * [David Holmes](https://openjdk.org/census#dholmes) (@dholmes-ora - **Reviewer**)


### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/15159/head:pull/15159` \
`$ git checkout pull/15159`

Update a local copy of the PR: \
`$ git checkout pull/15159` \
`$ git pull https://git.openjdk.org/jdk.git pull/15159/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 15159`

View PR using the GUI difftool: \
`$ git pr show -t 15159`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/15159.diff">https://git.openjdk.org/jdk/pull/15159.diff</a>

</details>


### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/15159#issuecomment-1666465733)

tstuefe commented:
Any takers?

kstefanj commented on test/hotspot/jtreg/runtime/os/TestHugePageDecisionsAtVMStartup.java:
```diff
@@ -0,0 +1,165 @@
+
+        if (!useLP) {
+            out.shouldContain("[info][pagesize] Large page support disabled");
+        }
+
+        else if (useLP && !useTHP &&
```
I would prefer skipping the empty lines here and below.
```suggestion
        } else if (useLP && !useTHP &&
```

kstefanj reviewed (approved):
Looks good, just one minor thing. 

I saw the comment about `LargePageSizeInBytes` and thought that it should be required away but realized that `ProcessTools.createJavaProcessBuilder()` doesn't add the passed in options.

Also did some manual runs on an aarch64 system where the default 512m pages are not configured but the 2m pages are. This seem to generate output which will be accepted by the test.

tstuefe commented:
Thanks @kstefanj!

Since we just changed the error warning texts (https://github.com/openjdk/jdk/pull/15158), I needed to change the error text somewhat.

I then retested on x64 with the following configurations:

- No Hugepages configured (but supported by kernel) and:
  - THP "never"
  - THP "always"
  - THP "advise"

- THP "never" and 2M default hugepage size and:
  - nonzero Hugepages enabled (2M)
  - nonzero Hugepages enabled (1G)
  - nonzero Hugepages enabled (2M, 1G)

- THP "never" and 1G default hugepage size (via kernel parameter) and nonzero 1G Hugepages
 ( this uncovered a small test bug where "buildSizeString()" had been wrong)

- Hugepages completely absent on a kernel built without hugepages (aarch64 SBC)
	(this uncovered another test bug where HugePageConfiguration would fall back to THP mode "unknown" which then confused the test. I removed that enum value.)

All issues have been fixed.

kstefanj commented on test/hotspot/jtreg/runtime/os/TestHugePageDecisionsAtVMStartup.java:
```diff
@@ -0,0 +1,163 @@
+        } else if (useLP && !useTHP &&
+                 (!configuration.supportsStaticHugePages() || !haveUsableStaticHugePages)) {
+            out.shouldContain(warningNoLP);
+        }
+
+        else if (useLP && useTHP && !configuration.supportsTHP()) {
```
```suggestion
        } else if (useLP && useTHP && !configuration.supportsTHP()) {
```

kstefanj commented on test/hotspot/jtreg/runtime/os/TestHugePageDecisionsAtVMStartup.java:
```diff
@@ -0,0 +1,163 @@
+
+        else if (useLP && useTHP && !configuration.supportsTHP()) {
+            out.shouldContain(warningNoTHP);
+        }
+
+        else if (useLP && !useTHP &&
```
```suggestion
        } else if (useLP && !useTHP &&
```

kstefanj commented on test/hotspot/jtreg/runtime/os/TestHugePageDecisionsAtVMStartup.java:
```diff
@@ -0,0 +1,163 @@
+            out.shouldContain("[info][pagesize] Using the default large page size: " + buildSizeString(configuration.getStaticDefaultHugePageSize()));
+            out.shouldContain("[info][pagesize] UseLargePages=1, UseTransparentHugePages=0");
+            out.shouldContain("[info][pagesize] Large page support enabled");
+        }
+
+        else if (useLP && useTHP && configuration.supportsTHP()) {
```
```suggestion
        } else if (useLP && useTHP && configuration.supportsTHP()) {
```

tstuefe commented:
Thanks @kstefanj !

Took over your suggestions, they make the code look cleaner.

tstuefe commented:
Friendly ping

tstuefe commented:
Friendly ping. Only need a second review.

dholmes-ora reviewed (approved):
Seems okay. Thanks

tstuefe commented:
Thank you, David!

/integrate
